src/Shared/CopyOnWriteDictionary.cs

```diff
@@ -34,25 +36,13 @@ namespace Microsoft.Build.Collections
     /// be run in a separate appdomain.
     /// </comment>
     [Serializable]
-    internal class CopyOnWriteDictionary<K, V> : IDictionary<K, V>, IDictionary where V : class
+    internal class CopyOnWriteDictionary<K, V> : IDictionary<K, V>, IDictionary, ISerializable where V : class
```

Comment:
This class is internal; can you just delete it in favor of the ImmutableDictionary? If you do this and use System.Collections.Immutable.ImmutableDictionary, you can also get rid of the tests.

Reply:
No, because of the serializability constraint. `ImmutableDictionary` isn't, so I can either encapsulate it and define the serialization here once, or push it to all classes that have a CopyOnWriteDictionary.